cpp/kiss_icp/core/VoxelHashMap.cpp

```diff
@@ -49,64 +49,54 @@ namespace kiss_icp {
 VoxelHashMap::Vector3dVectorTuple VoxelHashMap::GetCorrespondences(
     const Vector3dVector &points, double max_correspondance_distance) const {
     // Lambda Function to obtain the KNN of one point, maybe refactor
-    auto GetClosestNeighboor = [&](const Eigen::Vector3d &point) {
+    auto GetClosestNeighbour = [&](const Eigen::Vector3d &point) {
         auto kx = static_cast<int>(point[0] / voxel_size_);
         auto ky = static_cast<int>(point[1] / voxel_size_);
         auto kz = static_cast<int>(point[2] / voxel_size_);
-        std::vector<Voxel> voxels;
-        voxels.reserve(27);
+
+        Eigen::Vector3d closest_neighbour;
+        double closest_distance2 = std::numeric_limits<double>::max();
+
         for (int i = kx - 1; i < kx + 1 + 1; ++i) {
             for (int j = ky - 1; j < ky + 1 + 1; ++j) {
                 for (int k = kz - 1; k < kz + 1 + 1; ++k) {
-                    voxels.emplace_back(i, j, k);
-                }
-            }
-        }
-
-        using Vector3dVector = std::vector<Eigen::Vector3d>;
-        Vector3dVector neighboors;
-        neighboors.reserve(27 * max_points_per_voxel_);
-        std::for_each(voxels.cbegin(), voxels.cend(), [&](const auto &voxel) {
-            auto search = map_.find(voxel);
-            if (search != map_.end()) {
-                const auto &points = search->second.points;
-                if (!points.empty()) {
-                    for (const auto &point : points) {
-                        neighboors.emplace_back(point);
+                    auto search = map_.find({i, j, k});
```

Comment:
Would you mind being a bit more explicit (on the PR description, not the code) about the change you just made?

I kind of reverse engineered the new solution and I think is better, but if would be nice for the other reviewers to facilitate a bit the high level idea ;)

Reply:
I agree the readability is not great, I'll profile again the code this evening to see if there is some real improvements and give explanations